ground/src/main/java/com/google/android/ground/ui/map/gms/features/FeatureManager.kt

```diff
@@ -0,0 +1,117 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.ground.ui.map.gms.features
+
+import android.content.Context
+import com.google.android.gms.maps.GoogleMap
+import com.google.android.gms.maps.model.LatLng
+import com.google.android.ground.ui.map.Feature
+import dagger.hilt.android.qualifiers.ApplicationContext
+import javax.inject.Inject
+import timber.log.Timber
+
+/**
+ * Coordinates adding, removing, and updating of [Feature]s to the map. This abstracts access to
+ * both rendering of individual features, as well as clustering. This class is not thread-safe.
+ */
+class FeatureManager
+@Inject
+constructor(
+  @ApplicationContext private val context: Context,
+  private val pointRenderer: PointRenderer,
+  private val polygonRenderer: PolygonRenderer,
+  private val lineStringRenderer: LineStringRenderer
+) {
+  private val features = mutableSetOf<Feature>()
+  private val featuresByTag = mutableMapOf<Feature.Tag, Feature>()
+
+  private lateinit var map: GoogleMap
+  private lateinit var mapsItemManager: MapsItemManager
+  private lateinit var clusterManager: FeatureClusterManager
+  private lateinit var clusterRenderer: FeatureClusterRenderer
+
+  /**
+   * The camera's current zoom level. This must be set here since this impl can't access
+   * `map.cameraPosition` from off the main UI thread.
+   */
+  var zoom: Float
+    get() = clusterRenderer.zoom
+    set(value) {
+      clusterRenderer.zoom = value
+    }
+
+  /** Clears all managed state an binds to the provided [GoogleMap]. */
+  fun onMapReady(map: GoogleMap) {
+    features.clear()
+    featuresByTag.clear()
+    mapsItemManager = MapsItemManager(map, pointRenderer, polygonRenderer, lineStringRenderer)
+    clusterManager = FeatureClusterManager(context, map)
+    clusterRenderer = FeatureClusterRenderer(context, map, clusterManager, map.cameraPosition.zoom)
+    clusterRenderer.onClusterItemRendered = { mapsItemManager.setVisible(it, true) }
+    clusterRenderer.onClusterRendered = { mapsItemManager.setVisible(it, false) }
+    clusterManager.renderer = clusterRenderer
+    this.map = map
+  }
+
+  /**
+   * Updates the current set of features managed by the manager, adding and removing items from the
+   * map as needed to sync the map state with the provided collection.
+   */
+  fun setFeatures(updatedFeatures: Collection<Feature>) {
+    // remove stale
+    val removedOrChanged = features - updatedFeatures.toSet()
+    removedOrChanged.forEach(this::remove)
```

Comment:
Nit use trailing lambda or method reference consistently (we use a trailing lambda below).

Reply:
I'm on the fence as to which is more readable here, but I agree they should be consistent. Using method ref for now. Some valuable insight here: https://stackoverflow.com/a/24493905/9986466

Reply:
Same here. I don't really care which we use so long as it's (locally) consistent in the body of this function. I have a slight preference for method reference in this case since we're doing nothing other than transparently passing through a unary argument.

Reply:
+1, same preference here.